### PR TITLE
Anti feed system

### PR DESCRIPTION
**Pull request by Sheodar**
System is detailed in this issue https://github.com/dota2unofficial/12v12/issues/54

### CODE REVIEW

File: game/scripts/vscripts/addon_game_mode.lua [hunk 1]
```diff
@@ -153,16 +180,111 @@ function otherTeam(team)
...
+function UnitInSafeZone(unit , unitPosition)
+	local teamNumber = unit:GetTeamNumber()
+	local fountains = Entities:FindAllByClassname('ent_dota_fountain')
+	local allyFountainPosition
+	for i, focusFountain in pairs(fountains) do
+		if focusFountain:GetTeamNumber() == teamNumber then
+			allyFountainPosition = focusFountain:GetAbsOrigin()
+		end
+	end
+	if ((allyFountainPosition - unitPosition):Length2D()) <= TROLL_FEED_DISTANCE_FROM_FOUNTAIN_TRIGGER then
```
Comment (ark120202) [inline game/scripts/vscripts/addon_game_mode.lua]:
```suggestion
	return ((allyFountainPosition - unitPosition):Length2D()) <= TROLL_FEED_DISTANCE_FROM_FOUNTAIN_TRIGGER
```

File: game/scripts/vscripts/addon_game_mode.lua [hunk 2]
```diff
@@ -153,16 +180,111 @@ function otherTeam(team)
...
+		focusTableHeroes = _G.tableRadiantHeroes
+	elseif unitTeam == DOTA_TEAM_BADGUYS then
+		focusTableHeroes = _G.tableDireHeroes
+	end
+
+	for i, focusHero in pairs(focusTableHeroes) do
+		local unitKD = GetHeroKD(unit)
+		if unitKD > TROLL_FEED_RATIO_KD_TO_TRIGGER_MIN then
+			return false
+		elseif (GetHeroKD(focusHero) <= unitKD) and (not(unit == focusHero))then
```
Comment (ark120202) [inline game/scripts/vscripts/addon_game_mode.lua]:
```suggestion
		elseif GetHeroKD(focusHero) <= unitKD and unit ~= focusHero then
```

File: game/scripts/vscripts/addon_game_mode.lua [hunk 3]
```diff
@@ -466,7 +624,7 @@ function CMegaDotaGameMode:OnGameRulesStateChange(keys)
         }
 
         local fountains = Entities:FindAllByClassname('ent_dota_fountain')
-        -- Loop over all ents
+		--        -- Loop over all ents
```
Comment (ark120202) [inline game/scripts/vscripts/addon_game_mode.lua]:
```suggestion
		-- Loop over all ents
```

File: game/scripts/vscripts/addon_game_mode.lua [hunk 4]
```diff
@@ -153,16 +180,111 @@ function otherTeam(team)
...
+	return true
+end
+
+function CMegaDotaGameMode:OnHeroPicked(event)
+	local hero = EntIndexToHScript(event.heroindex)
+
+	if hero then
+		if hero:GetTeamNumber() == DOTA_TEAM_GOODGUYS then
+			table.insert(_G.tableRadiantHeroes, hero)
+			_G.tableRadiantHeroes[#tableRadiantHeroes].GetHeroName = function()
```
Comment (ark120202) [inline game/scripts/vscripts/addon_game_mode.lua]: At the very least that's pretty confusing. If you want to mutate hero, why not just do it in-place? And what's the point of overriding it at all? Where is `hero_name` defined?

File: game/scripts/vscripts/addon_game_mode.lua [hunk 5]
```diff
@@ -153,16 +180,111 @@ function otherTeam(team)
...
+			return false
+		end
+	end
+	return true
+end
+
+function CMegaDotaGameMode:OnHeroPicked(event)
+	local hero = EntIndexToHScript(event.heroindex)
+
+	if hero then
```
Comment (ark120202) [inline game/scripts/vscripts/addon_game_mode.lua]: You can use early return, i.e.

```lua
local hero = EntIndexToHScript(event.heroindex)
if not hero then return end

if hero:GetTeamNumber() == DOTA_TEAM_GOODGUYS then
...
```

**Comment by darklordabc** [comment]
Can we merge this @ark120202 ?

File: game/scripts/vscripts/addon_game_mode.lua [hunk 6]
```diff
@@ -223,20 +219,15 @@ end
 
 function CMegaDotaGameMode:OnHeroPicked(event)
 	local hero = EntIndexToHScript(event.heroindex)
+	if not hero then return end
 
 	if hero then
```
Comment (ark120202) [inline game/scripts/vscripts/addon_game_mode.lua]: Using early return makes this condition unnecessary

**Pull request merged**